tasks/generate_bdi_data.py

```diff
@@ -0,0 +1,302 @@
+import tempfile
+import shutil
+import os
+import math
+from cumulusci.tasks.salesforce import BaseSalesforceApiTask
+from cumulusci.tasks.bulkdata import LoadData
+from cumulusci.core.utils import ordered_yaml_load
+from cumulusci.utils import convert_to_snake_case
+from cumulusci.core.config import TaskConfig
+from datetime import date
+from datetime import timedelta
+from sqlalchemy import create_engine
+from sqlalchemy import Column
+from sqlalchemy import MetaData
+from sqlalchemy import Integer
+from sqlalchemy import Table
+from sqlalchemy import Unicode
+from sqlalchemy.ext.automap import automap_base
+from sqlalchemy.orm import create_session
+
+START_DATE = date(2019, 1, 1)
+
+
+class GenerateBDIData(BaseSalesforceApiTask):
+    def _run_task(self):
+        self.tempdir = tempfile.mkdtemp()
+        try:
+            sqlite_path = os.path.join(self.tempdir, "generated_data.db")
+            url = "sqlite:///" + sqlite_path
+            batch_size = math.floor(self.options["num_records"] / 10)
+            self.generate_data(url, self.options["mapping"], batch_size)
+            subtask_config = TaskConfig(
+                {"options": {"database_url": url, "mapping": self.options["mapping"]}}
+            )
+            subtask = LoadData(
+                project_config=self.project_config,
+                task_config=subtask_config,
+                org_config=self.org_config,
+                flow=self.flow,
+                name=self.name,
+                stepnum=self.stepnum,
+            )
+            subtask()
+        finally:
+            shutil.rmtree(self.tempdir)
+
+    def generate_data(self, db_url, mapping_file_path, batch_size):
+        with open(mapping_file_path, "r") as f:
+            mappings = ordered_yaml_load(f)
+
+        self.session, base = init_db(db_url, mappings)
+        self.make_all_records(batch_size, base)
+        self.generate_bdi_denormalized_table(2)
+        self.session.commit()
+
+    def make_opportunity(self, amount, date, paid, payment_amount, **kw):
+        opp = self.Opportunity(
+            amount=amount, stage_name="Prospecting", close_date=date, **kw
+        )
+        self.session.add(opp)
+        self.session.flush()
+        if payment_amount:
+            payment = self.Payment(
+                npe01__opportunity__c=opp.id,
+                amount=payment_amount,
+                payment_date=date,
+                paid=paid,
+            )
+            self.session.add(payment)
+
+    def make_records(
+        self, model, name, key_field, start, end, amount, paid, payment_amount
+    ):
+        date = START_DATE
+        for i in range(start + 1, end + 1):
+            parent = model(name=name + " " + str(i))
+            print(name + " " + str(i))
```

Comment:
If this is for temporary debugging, take it out; if it's something you want to see in MetaCI, use self.logger